x/mongo/driver/topology/connection_test.go

```diff
@@ -1101,6 +1101,67 @@ func (tcl *testCancellationListener) assertCalledOnce(t *testing.T) {
 	assert.Equal(t, 1, tcl.numStopListening, "expected StopListening to be called once, got %d", tcl.numListen)
 }
 
+type testContext struct {
+	context.Context
+	deadline time.Time
+}
+
+func (tc *testContext) Deadline() (time.Time, bool) {
+	return tc.deadline, false
+}
+
+func TestConnectionError(t *testing.T) {
+	t.Parallel()
+
+	t.Run("EOF", func(t *testing.T) {
+		t.Parallel()
+
+		addr := bootstrapConnections(t, 1, func(nc net.Conn) {
+			_ = nc.Close()
+		})
+
+		p := newPool(
+			poolConfig{Address: address.Address(addr.String())},
+		)
+		defer p.close(context.Background())
+		err := p.ready()
+		require.NoError(t, err, "unexpected close error")
+
+		conn, err := p.checkOut(context.Background())
+		require.NoError(t, err, "unexpected checkOut error")
+
+		_, err = conn.readWireMessage(context.Background())
+		assert.ErrorContains(t, err, "connection closed unexpectedly by the other side: EOF")
+	})
+	t.Run("timeout", func(t *testing.T) {
+		t.Parallel()
+
+		timeout := 10 * time.Millisecond
+
+		addr := bootstrapConnections(t, 1, func(nc net.Conn) {
+			time.Sleep(timeout * 2)
+			_ = nc.Close()
+		})
+
+		p := newPool(
+			poolConfig{Address: address.Address(addr.String())},
+		)
+		defer p.close(context.Background())
+		err := p.ready()
+		require.NoError(t, err)
+
+		conn, err := p.checkOut(context.Background())
+		require.NoError(t, err)
+
+		ctx := &testContext{
+			Context:  context.Background(),
+			deadline: time.Now().Add(timeout),
+		}
+		_, err = conn.readWireMessage(ctx)
```

Comment:
We don't need to create a custom test context:
```suggestion
		ctx, cancel := context.WithTimeout(context.Background(), timeout)
		_, err = conn.readWireMessage(ctx)
		cancel()
```

Reply:
The custom test context bypasses the `contextDeadlineUsed` check in the `transformNetworkError()` with a false value, so it tests the logic in `ConnectionError.Error()`. Instead, the context returned by `WithTimeout()` triggers L344 in "connection.go" which has been tested in `TestBackgroundRead`.

Reply:
Ah I see, good call. Thanks! Can we add a compile-time check to assert that testContext implements context.Context?

Reply:
I don't think the assertion is necessary as it embeds a context and is also passed to a context parameter.